### PR TITLE
add send/2 override for multiple pids

### PR DESCRIPTION
**Pull request by devsnek**
From discussion in https://erlangforums.com/t/send-optimization-for-sending-to-many-processes/2200

TL;DR: sending a single message to many many pids by looping over the pids is not great because the pids are almost certainly guaranteed to be grouped into a much smaller number of nodes. By passing a list of the pids to send/2, we can group them by node, and only send one message to each node, which then distributes the message to its processes.

I poorly benchmarked this on my home network with a few laptops and found that it significantly reduces the time spent in `send`, with the list variant being around 10x faster. (testing script: https://gist.github.com/devsnek/d5d1b55eade3a3d5aad07fd3675f3a22)

TODO:
- [ ] does this make sense?
- [ ] tests
- [ ] docs

### CODE REVIEW

**Comment by rickard-green** [comment]
Over the years we've had plans to allow a list of destination identifiers in send operations in order to optimize multicast scenarios. We haven't seen any real use cases for that though, but it seems that you got one. Since there is a use case for it I think we should go forward with it.

A major issue with the implementation in this draft is the distribution of messages to processes on the receiving node. Since there is no limit on the amount of receivers on the receiving node, this code (in the `DOP_SEND` case in `erts_net_message()`) might hog the scheduler thread, that handles the message, for a long time which is not acceptable. This operation needs to yield (i.e., give control back the the scheduler so it can perform other tasks that it needs to perform) if the amount of receivers is large. Yielding here is not trivial though.

There are multiple ways to implement the yielding, but the following is the one that I think probably is the best and probably the easiest to implement:

`erts_net_message()` needs to save current state, return to the caller and inform it that it needs to continue with the current operation later, and also complete this operation before handling more incoming data. `erts_net_message()` is called in two different contexts depending on how the distribution was started:

1. From the `erlang:dist_ctrl_put_data/2` BIF when the distribution controller is a process (e.g. distribution over TLS). This is relatively straight forward by bumping all reductions and schedule continued excecution in a continuation BIF using the trap functionality.
2. From `driver_output_binary()`/`driver_output_binary2()` when the distribution controller is a port. (e.g. the default TCP distribution). This is a bit tricker, but doable. When yielding, one have to save the state in port specific data (`ERTS_PRTSD_*`). If one is called again in `driver_output_binary()`/`driver_output_binary2()`, continue with the ongoing `DOP_SEND` operation and queue the data passed to `driver_output_binary()`/`driver_output_binary2()` for later execution. On can however not be sure that there will be another call to `driver_output_binary()`/`driver_output_binary2()`, so one also need to set up a timer (using zero timeout) and continue handling the data in the timer callback if it has not already been handled via `driver_output_binary()`/`driver_output_binary2()`.

Updating the bootstrap is not needed in this case, so please remove those changes. Updating the preloaded beam files are needed for you to be able to test your code. We can however for security reasons not include binary files into our daily builds since we cannot inspect them, but I'll rebuild them before doing that.

You've changed the `erlang:send()` BIF from being implemented in C to being implemented in Erlang. This will cause an overhead on all uses of `erlang:send()` (not large but still) regardless of whether a multi-send is done or not. The vast majority of sends will not be a multi-send. To minimize the overhead for ordinary sends you want to keep it as a C-function and instead trap out to Erlang code in `erts_internal` when necessary (using the `TRAP` macros in `bif.h`). This way also other send operations such as `!`, `send_nosuspend`, etc can be able to handle multi-send.

**Comment by michalmuskala** [comment]
I think there's a second opportunity for optimisation. Once all the data is flattened and we know it's "well behaved" in a sequential chunk of memory (after the first copy), a more efficient copying routine could be used - `copy_shallow`, rather than `copy` - in the context of an ETS table this gives considerable performance improvements, so I'd imagine it might be similar here.
So there might be benefits even in a single-node scenario.

I'd also consider using a different function name - perhaps `send_many` or `send_all`. The `send` function is already very complicated with so many variants, caveats, and remarks in documentation. I think it would make code comprehension much simpler to have just a separate function to do this, arguably quite different, operation.

**Comment by devsnek** [comment]
@rickard-green thanks for taking a look at this. i think the changes you've described make sense but are a bit beyond the time/willingness i have to muck about in this codebase. if someone is willing to figure out the receiver suspending code i can update the rest, otherwise i think this will be stalled for the moment.

maybe as an alternative, and i'm not too familiar with beam internals so this could be a very dumb idea, but there is a `net_kernel` process that runs when distribution is enabled. could we theoretically just send the message and pid list to that, and handle the fanout in erlang code, so that we get yields for free? i imagine it would be slightly slower than the C but the overall wins from batching would dwarf it.

@michalmuskala this sounds reasonable. ultimately i don't think other things like `!` really need to be able to handle this use case, so separating them is fine.

**Comment by rickard-green** [comment]
@michalmuskala 
> I think there's a second opportunity for optimisation. Once all the data is flattened and we know it's "well behaved" in a sequential chunk of memory (after the first copy), a more efficient copying routine could be used - copy_shallow, rather than copy - in the https://github.com/erlang/otp/pull/6272 this gives considerable performance improvements, so I'd imagine it might be similar here.

Yes, it could be decoded first and then (shallow) copied on internal term format to receivers instead of distributing the message on external term format to all receivers.

> So there might be benefits even in a single-node scenario.

Yes.

@devsnek 

> i think the changes you've described make sense but are a bit beyond the time/willingness i have to muck about in this codebase. if someone is willing to figure out the receiver suspending code i can update the rest, otherwise i think this will be stalled for the moment.

I'll leave the PR open for a while to see if someone is interested in picking it up.

> maybe as an alternative, and i'm not too familiar with beam internals so this could be a very dumb idea, but there is a `net_kernel` process that runs when distribution is enabled. could we theoretically just send the message and pid list to that, and handle the fanout in erlang code, so that we get yields for free? i imagine it would be slightly slower than the C but the overall wins from batching would dwarf it.

`net_kernel` performs a lot of important tasks and we don't want to bother it with this. Another process could however be created for this purpose. Just dispatching such messages over another process would however break the [signal order guarantee](https://www.erlang.org/doc/reference_manual/processes.html#signal-delivery) (two signals received by a process from the same sender must be received in the same order as sent), so this would have to be solved which complicates this a lot unless you are willing to lose performance in the ordinary send case as well (which we are not).

> ultimately i don't think other things like ! really need to be able to handle this use case, so separating them is fine.

I don't agree.

**Comment by josevalim** [comment]
Could this also be used to provide additional guarantees? Today if you do this:

    [X ! Message || X <- Pids],

There is always a chance the process will be killed between messages. But perhaps this could provide additional guarantees in relation to _sending_, for example, all pids that belong to a given node are guaranteed to be sent the message.

It is also worth testing and documenting as well what happens when the same PID appears multiple times on the list.

**Comment by devsnek** [comment]
I updated send to use a trap back to erts_internal. I still have no idea how to implement the requested changes for `erts_net_message` :sweat_smile: 

> I think it would make code comprehension much simpler to have just a separate function to do this, arguably quite different, operation.

> I don't agree.

I do wonder what the expected behavior of nosuspend/noconnect would be with this. For example if one of the nodes you're sending to isn't connected yet but others are, would only some of the messages get sent, up until it processes the node that isn't connected?

**Comment by NelsonVides** [comment]
Hi guys, I just came across this proposal and I am interested in the change too. I'm trying to wrap my head around the current state of the PR, so let me paraphrase: the main point missing in the current PR is yielding the scheduler, to avoid hogging it when the list of receivers is too long or the payload to big and the operation takes too long.

Also paraphrasing the current implementation, it was agreed that we want the same API for `send` and `multisend` (which in turn applies to `!` too), and in the current PR this is implemented by having the `send` bif recognise that `to` is a list and then installing a trap to the `multisend_trap` function pointer, which will call erlang code to group the pids by node, and then calls `erts_internal:multisend_bif/2` which is implemented in C in `erts_internal_multisend_bif_2` (and might in turn put more traps to itself until is done). Is my understanding correct here? In this case, all is missing from [Rickard comment](https://github.com/erlang/otp/pull/6638#issuecomment-1405557877) is point 2, right?

Other points are tests and documentation, specially interesting are the guarantees as mentioned by Jose and the expected behaviour of nosuspend/noconnect as mentioned by the OP.

**Comment by devsnek** [comment]
sounds about right.

**Comment by NelsonVides** [comment]
> sounds about right.

Tried to do a bit of progress on this in #8072 🙂
